notebooks/query-dependent link analysis.ipynb

```diff
@@ -0,0 +1,1064 @@
+    "    node_names.append(v['properties']['name'])  # append cid id and name    \n",
+    "node_names_dict = dict(zip(node_names[::2], node_names[1::2]))  # turn list into dict \n",
+    "g_degree = {v: node_names_dict.get(k, v) for k, v in g_degree.items()}  # replace cid with node name \n",
+    "list(g_degree.items())[:10][:10]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "fc683c5a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Closeness centrality: closeness to other nodes\n",
+    "g_closeness = nx.closeness_centrality(g)\n",
+    "g_closeness = dict(sorted(g_closeness.items(), key=operator.itemgetter(1),reverse=True))\n",
+    "\n",
+    "node_names = []\n",
+    "for k,v in g.nodes(data=True):\n",
+    "    node_names.append(k)\n",
+    "    node_names.append(v['properties']['name'])  # append cid id and name    \n",
+    "node_names_dict = dict(zip(node_names[::2], node_names[1::2]))  # turn list into dict \n",
+    "g_closeness = {v: node_names_dict.get(k, v) for k, v in g_closeness.items()}  # replace cid with node name \n",
+    "list(g_closeness.items())[:10][:10]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "21f68319",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Eigenvector centrality \n",
+    "g_di = nx.DiGraph(g)\n",
+    "g_eigen = nx.eigenvector_centrality(g_di)\n",
+    "g_eigen = dict(sorted(g_eigen.items(), key=operator.itemgetter(1),reverse=True))\n",
+    "\n",
+    "node_names = []\n",
+    "for k,v in g.nodes(data=True):\n",
+    "    node_names.append(k)\n",
+    "    node_names.append(v['properties']['name']) # append cid id and name    \n",
+    "node_names_dict = dict(zip(node_names[::2], node_names[1::2]))  # turn list into dict \n",
+    "g_eigen = {v: node_names_dict.get(k, v) for k, v in g_eigen.items()}  # replace cid with node name \n",
+    "list(g_eigen.items())[:10][:10]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "14b924d5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# VoteRank\n",
+    "g_voterank = nx.voterank(g)\n",
+    "\n",
+    "node_names = []\n",
+    "for k,v in g.nodes(data=True):\n",
+    "    node_names.append(k)\n",
+    "    node_names.append(v['properties']['name']) # append cid id and name   \n",
+    "node_names_dict = dict(zip(node_names[::2], node_names[1::2]))  # turn list into dict \n",
+    "g_voterank = [node_names_dict.get(item,item)  for item in g_voterank]\n",
+    "g_voterank[:10]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "50bbf9da",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Which nodes have user_movement in or out? 186\n",
+    "user_movement_nodes = []\n",
+    "\n",
+    "for node1, node2, data in g.edges(data=True):\n",
+    "    if data['type']=='USER_MOVEMENT':\n",
+    "            user_movement_nodes.append(node1)\n",
```

Comment:
Can this all be rewritten as:

user_movement_nodes = [node1 for node1,node2,data in g.edges(data=True) if data['type'] == 'USER_MOVEMENT']

Reply:
Thanks again, refactored in commit cbeea18